### ISSUE TITLE
Jetty 12.0 Handle HEAD requests in Handler

### ISSUE THREAD

**Issue by poutsma** [eclipse/jetty.project#9953]
Hi,

I work on the Spring Framework, where we are in the process of upgrading to Jetty 12 as part of our 6.1 branch. As part of this migration, I have run into an issue.

This issue occurs when an asynchronous `WriteListener` does not write the amount of bytes specified in `Content-Length`. This is fine for typical requests, but not for `HEAD` requests, given [Section 14.13 of RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13):

> The Content-Length entity-header field indicates the size of the entity-body [...] sent to the recipient or, **in the case of the HEAD method, the size of the entity-body that would have been sent had the request been a GET**.

The following Servlet can be used to reproduce this issue:

```java
@WebServlet(urlPatterns={"/"}, asyncSupported=true)
public class AsyncServlet extends GenericServlet {
	@Override
	public void service(ServletRequest req, ServletResponse res) throws IOException {
		HttpServletRequest request = (HttpServletRequest) req;
		if (request.getMethod().equals("GET") || request.getMethod().equals("HEAD")) {
			AsyncContext context = req.startAsync();
			res.getOutputStream().setWriteListener(new HeadWriteListener(context));
		}
		else {
			HttpServletResponse response = (HttpServletResponse) res;
			response.sendError(404);
		}
	}

	private static class HeadWriteListener implements WriteListener {
		private final AsyncContext context;

		public HeadWriteListener(AsyncContext context) {
			this.context = context;
		}

		@Override
		public void onWritePossible() throws IOException {
			HttpServletRequest req = (HttpServletRequest) this.context.getRequest();
			HttpServletResponse resp = (HttpServletResponse) this.context.getResponse();
			resp.setContentType("text/plain");
			resp.setContentLength(9);
			if (req.getMethod().equals("GET")) {
				try (ServletOutputStream outputStream = resp.getOutputStream()) {
					outputStream.write("FooBarBaz".getBytes(UTF_8));
				}
			}
			this.context.complete();
		}

		@Override
		public void onError(Throwable t) {
			t.printStackTrace();
		}
	}
}
```

Running this servlet on Jetty 12.0 beta 2 results in the following exception in `HttpChannelState`:

```
java.io.IOException: written 0 < 9 content-length
	at org.eclipse.jetty.server.internal.HttpChannelState$ChannelResponse.write(HttpChannelState.java:1208)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.channelWrite(HttpOutput.java:215)
	at org.eclipse.jetty.ee10.servlet.HttpOutput.complete(HttpOutput.java:447)
	at org.eclipse.jetty.ee10.servlet.ServletContextResponse.completeOutput(ServletContextResponse.java:209)
	at org.eclipse.jetty.ee10.servlet.ServletChannel.handle(ServletChannel.java:625)
	at org.eclipse.jetty.ee10.servlet.ServletHandler.handle(ServletHandler.java:458)
	at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:798)
	at org.eclipse.jetty.server.Handler$Wrapper.handle(Handler.java:611)
	at org.eclipse.jetty.server.Server.handle(Server.java:177)
	at org.eclipse.jetty.server.internal.HttpChannelState$HandlerInvoker.run(HttpChannelState.java:617)
	at org.eclipse.jetty.server.internal.HttpConnection.onFillable(HttpConnection.java:480)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:473)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:436)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:288)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.produce(AdaptiveExecutionStrategy.java:196)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
**Jetty version(s)**
Jetty 12.0 beta2

**Java version/vendor** `(use: java -version)`

openjdk version "17.0.7" 2023-04-18 LTS

**Comment by gregw** [comment]
Typically the container ignores content written, so you don't need to explicitly handle head like this servlet does.    However, I'll look to see if we can avoid that length written check for head methods when only 0 bytes have been written... stand by...

**Comment by gregw** [comment]
See PR #9957 This is not actually related to Aysnc, but to any handler that self handles HEAD.

**Comment by joakime** [comment]
Merged in PR #9957
